File: backend/utils/timeoutManager.js
/**
 * Gestionnaire de timeouts maîtrisés
 * Timeouts adaptatifs selon le type d'opération
 */
class TimeoutManager {
  constructor() {
    this.timeouts = {
      database: 10000, // 10 secondes
      cache: 2000, // 2 secondes
      externalAPI: 5000, // 5 secondes
      fileUpload: 300000, // 5 minutes
      fileDownload: 60000, // 1 minute
      default: 30000, // 30 secondes
    };
  }

  /**
   * Créer un timeout avec Promise
   */
  createTimeout(promise, type = 'default', customTimeout = null) {
    const timeout = customTimeout || this.timeouts[type] || this.timeouts.default;

    return Promise.race([
      promise,
      new Promise((_, reject) => {
        setTimeout(() => {
          reject(new Error(`Operation timed out after ${timeout}ms`));
        }, timeout);
      }),
    ]);
  }

  /**
   * Wrapper pour opérations avec timeout
   */
  async withTimeout(fn, type = 'default', customTimeout = null) {
    return await this.createTimeout(
      Promise.resolve(fn()),
      type,
      customTimeout
    );
  }

  /**
   * Timeout adaptatif selon la charge
   */
  getAdaptiveTimeout(baseType, loadFactor = 1) {
    const baseTimeout = this.timeouts[baseType] || this.timeouts.default;
    
    // Augmenter le timeout si charge élevée
    if (loadFactor > 1.5) {
      return baseTimeout * 1.5;
    }
    
    return baseTimeout;
  }
}

module.exports = new TimeoutManager();

File: mobile-app/lib/utils/timeout_manager.dart
/// Gestionnaire de timeouts maîtrisés
/// Timeouts adaptatifs selon le type d'opération
import 'dart:async';

class TimeoutException implements Exception {
  final String message;
  final Duration timeout;

  TimeoutException(this.message, this.timeout);

  @override
  String toString() => message;
}

class TimeoutManager {
  static final TimeoutManager _instance = TimeoutManager._internal();
  factory TimeoutManager() => _instance;
  TimeoutManager._internal();

  final Map<String, Duration> _timeouts = {
    'database': const Duration(seconds: 10),
    'cache': const Duration(seconds: 2),
    'externalAPI': const Duration(seconds: 5),
    'fileUpload': const Duration(minutes: 5),
    'fileDownload': const Duration(minutes: 1),
    'default': const Duration(seconds: 30),
  };

  /// Créer un timeout avec Future
  Future<T> createTimeout<T>(
    Future<T> Function() promise,
    String type, {
    Duration? customTimeout,
  }) async {
    final timeout = customTimeout ?? _timeouts[type] ?? _timeouts['default']!;

    return await promise().timeout(
      timeout,
      onTimeout: () {
        throw TimeoutException('Operation timed out after ${timeout.inSeconds}s', timeout);
      },
    );
  }

  /// Wrapper pour opérations avec timeout
  Future<T> withTimeout<T>(
    Future<T> Function() fn,
    String type, {
    Duration? customTimeout,
  }) async {
    return await createTimeout(fn, type, customTimeout: customTimeout);
  }

  /// Timeout adaptatif selon la charge
  Duration getAdaptiveTimeout(String baseType, {double loadFactor = 1.0}) {
    final baseTimeout = _timeouts[baseType] ?? _timeouts['default']!;

    // Augmenter le timeout si charge élevée
    if (loadFactor > 1.5) {
      return Duration(
        milliseconds: (baseTimeout.inMilliseconds * 1.5).toInt(),
      );
    }

    return baseTimeout;
  }
}

